File: R/st_univariato_alpha.R
# GRID alpha — ST univariata (generazione per alpha, stima, diagnostiche UNIF + inversione)
# Output:
#   - common/tables/st_uni_A_grid_R.csv
#   - common/logs/st_uni_A_grid_times_R.csv

library(sn)

# Griglia di asimmetria alpha su cui simulare e stimare; n osservazioni per ciascun alpha.
# Parametri di generazione: xi (posizione), omega (scala), nu_true (gradi di libertà).
alpha_grid <- c(-4.0, -1.2, -0.2, 0.0, 0.2, 1.2, 4.0)
n       <- 2000
xi      <- 0
omega   <- 1
nu_true <- 6   # momenti fino al 4° ben definiti

# Soglia semantica per nu: garantisce almeno varianza esistente (≥2.05).
# Per pretendere anche kurtosi (4° momento), imposta ST_SEM_NU_MIN=4.05 nell'ambiente.
SEM_NU_MIN <- suppressWarnings(as.numeric(Sys.getenv("ST_SEM_NU_MIN", unset = "2.05")))
if (!is.finite(SEM_NU_MIN) || SEM_NU_MIN <= 0) SEM_NU_MIN <- 2.05

# Bound inferiori in scala log per omega e nu usati nei vincoli di ottimizzazione.
LOG_OM_MIN <- log(1e-8)
LOG_NU_MIN <- log(SEM_NU_MIN)

# Controlli per L-BFGS-B: iterazioni massime, factr (precisione), pgtol (gradiente).
LBFGSB_CTRL <- list(maxit = 5000, factr = 1e7, pgtol = 1e-8)

# Directory di output: dati, figure, tabelle, log. Le crea se non esistono.
dirs <- list(
  data = "common/data",
  fig  = "common/fig",
  tab  = "common/tables",
  log  = "common/logs"
)
for (d in dirs) dir.create(d, recursive = TRUE, showWarnings = FALSE)

# funzioni di timing
tic  <- function() proc.time()[["elapsed"]]
tock <- function(t0) proc.time()[["elapsed"]] - t0

# NLL per ST con penalizzazione dura fuori dai vincoli.
#     Parametrizzazione: (xi, log(omega), alpha, log(nu)).
#     - Controlli di finitezza.
#     - Guard-rails: omega>0, nu≥soglia semantica.
#     - try su densità: se fallisce o produce non-finiti/non-positivi, restituisce 1e50.
#     Ritorna la somma negativa dei log-densità (valore da minimizzare).
# parametri: (xi, log(omega), alpha, log(nu))
nll_st_par <- function(par, y) {
  xi  <- par[1]; lom <- par[2]; al <- par[3]; lnu <- par[4]
  if (!is.finite(lom) || !is.finite(lnu)) return(1e50)
  om <- exp(lom); nv <- exp(lnu)
  if (!is.finite(om) || om <= 0)          return(1e50)
  if (!is.finite(nv) || nv <  SEM_NU_MIN) return(1e50)

  dens <- suppressWarnings(try(dst(y, xi = xi, omega = om, alpha = al, nu = nv), silent = TRUE))
  if (inherits(dens, "try-error") || any(!is.finite(dens)) || any(dens <= 0)) return(1e50)
  -sum(log(dens))
}

# Loop su alpha
# Collezioni per accumulare risultati di stima e tempi per ogni alpha.
all_fit   <- list()
all_times <- list()

for (a in alpha_grid) {
  # (A) Dati
  # Seed deterministico che dipende da alpha (così ogni alpha produce sempre lo stesso dataset).
  set.seed(20250303 + as.integer(round((a + 10)*1000)))
  t0 <- tic()
  y  <- rst(n, xi = xi, omega = omega, alpha = a, nu = nu_true)
  t_gen <- tock(t0)

  # Salva il dataset simulato per questo alpha in un file separato (nome con alpha formattato).
  fdata <- sprintf("%s/st_uni_A_alpha_%s.csv", dirs$data, gsub("\\.", "p", sprintf("%.1f", a)))
  write.csv(data.frame(y = y), fdata, row.names = FALSE)

  # (B) Stima ML (xi, log(omega), alpha, log(nu)) con vincoli
  # Punto iniziale per alpha: 0 se a=0, altrimenti segno(a) e ampiezza limitata tra 0.1 e 2.
  #     Start per log(nu): almeno soglia+0.5 o nu_true, per stabilità numerica.
  start_alpha <- if (a == 0) 0 else sign(a) * max(0.1, min(2, abs(a)))
  start <- c(median(y), log(IQR(y)/1.349), start_alpha, log(max(SEM_NU_MIN + 0.5, nu_true)))

  # Ottimizzazione L-BFGS-B della NLL con vincoli su log(omega) e log(nu).
  t0 <- tic()
  fit <- optim(
    par     = start,
    fn      = function(p) nll_st_par(p, y),
    method  = "L-BFGS-B",
    lower   = c(-Inf, LOG_OM_MIN, -Inf, LOG_NU_MIN),
    control = LBFGSB_CTRL
  )
  t_fit <- tock(t0)

  # Estrae stime sullo spazio originale per omega e nu; log-likelihood massimo.
  parhat <- c(
    xi    = fit$par[1],
    omega = exp(fit$par[2]),
    alpha = fit$par[3],
    nu    = exp(fit$par[4])
  )
  llhat  <- -fit$value

  # (C) Diagnostiche sintetiche: inversione CDF↔PPF (core + code) e UNIF (KS)
  # Griglia di probabilità: “core” (0.001–0.999) + “extreme tails” (1e-6,1e-5,1e-4, ...).
  p_core <- seq(0.001, 0.999, by = 0.001)
  p_ext  <- c(1e-6, 1e-5, 1e-4, 1 - 1e-4, 1 - 1e-5, 1 - 1e-6)
  p_grid <- c(p_ext[1:3], p_core, p_ext[4:6])

  # Verifica di inversione: qst(p) seguito da pst(q). Calcola max errore assoluto
  # separatamente per core ed estremi.
  q_inv  <- suppressWarnings(qst(p_grid, xi = parhat["xi"], omega = parhat["omega"],
                                 alpha = parhat["alpha"], nu = parhat["nu"]))
  p_back <- suppressWarnings(pst(q_inv,  xi = parhat["xi"], omega = parhat["omega"],
                                 alpha = parhat["alpha"], nu = parhat["nu"]))
  inv_core <- max(abs(p_back[4:(length(p_back)-3)] - p_core))
  inv_ext  <- max(abs(p_back[c(1:3, (length(p_back)-2):length(p_back))] -
                       c(p_ext[1:3], p_ext[4:6])))

  # Trasformazione PIT (u = F(y; theta hat)); KS contro Uniforme(0,1) come check di adeguatezza.
  u    <- suppressWarnings(pst(y, xi = parhat["xi"], omega = parhat["omega"],
                               alpha = parhat["alpha"], nu = parhat["nu"]))
  ks_p <- suppressWarnings(ks.test(u, "punif")$p.value)  # UNIF: KS su U(0,1)

  # (D) Accumulo risultati
  # Raccoglie misure chiave per questo alpha: stime, loglik, errori d’inversione,
  # p-value KS, n e soglia nu usata; e logga i tempi di generazione e stima.
  all_fit[[length(all_fit)+1]] <- data.frame(
    engine="R", alpha_true=a,
    xi_hat=parhat["xi"], omega_hat=parhat["omega"],
    alpha_hat=parhat["alpha"], nu_hat=parhat["nu"],
    loglik=llhat, n=n,
    inv_core_maxerr=inv_core, inv_ext_maxerr=inv_ext, ks_pvalue=ks_p,
    nu_sem_threshold = SEM_NU_MIN
  )
  all_times[[length(all_times)+1]] <- data.frame(
    alpha_true=a, step="data_rst_generate", seconds=t_gen   # allineato al merge
  )
  all_times[[length(all_times)+1]] <- data.frame(
    alpha_true=a, step="fit_ML_LBFGSB", seconds=t_fit       # allineato al merge
  )
}

# Scrittura output aggregati 
fit_out   <- do.call(rbind, all_fit)
times_out <- do.call(rbind, all_times)
write.csv(fit_out,   file.path(dirs$tab, "st_uni_A_grid_R.csv"),           row.names=FALSE)
write.csv(times_out, file.path(dirs$log, "st_uni_A_grid_times_R.csv"),     row.names=FALSE)

# Messaggio di riepilogo a console con percorso dei file prodotti.
cat("✓ ST univariata — GRID alpha: scritto\n",
    "  • Tabelle :", file.path(dirs$tab, "st_uni_A_grid_R.csv"), "\n",
    "  • Tempi   :", file.path(dirs$log, "st_uni_A_grid_times_R.csv"), "\n", sep = "")

File: Python/st_univariato_alpha.py
# GRID alpha — ST univariata: fit, inversione CDF↔PPF, UNIF, timing
# Dipendenze: numpy, pandas, scipy, skewt-scipy

import os, time
import numpy as np
import pandas as pd
from scipy import stats

# import skewt-scipy
try:
    from skewt_scipy.skewt import skewt
except Exception as e:
    raise ImportError(
        "Impossibile importare 'skewt' dal pacchetto 'skewt-scipy'. "
        "Installa/aggiorna con: pip install -U skewt-scipy"
    ) from e

# Griglia di valori di alpha per cui leggere i dataset e fare il fit.
alpha_grid = [-4.0, -1.2, -0.2, 0.0, 0.2, 1.2, 4.0]  # come SN

# Directory base per dati, tabelle e log (coerenti con gli script R).
DATA_DIR = "common/data"
TAB_DIR  = "common/tables"
LOG_DIR  = "common/logs"

# Crea le cartelle per tabelle e log se non esistono già.
os.makedirs(TAB_DIR, exist_ok=True)
os.makedirs(LOG_DIR, exist_ok=True)

# Percorsi output complessivi (tabella risultati e tempi).
OUT_TAB = os.path.join(TAB_DIR, "st_uni_A_grid_Py.csv")
OUT_TIM = os.path.join(LOG_DIR, "st_uni_A_grid_times_Py.csv")

# Costruisce il percorso del file dati associato a un certo alpha.
# Esempio: alpha=-1.2 → "st_uni_A_alpha_-1.2" ma con il punto sostituito da "p".
def data_path_for_alpha(a: float) -> str:
    tag = str(round(a, 1)).replace(".", "p")
    return os.path.join(DATA_DIR, f"st_uni_A_alpha_{tag}.csv")

# Legge la colonna 'y' dal CSV corrispondente a un certo alpha e la restituisce come array NumPy.
def read_y(a: float) -> np.ndarray:
    fpath = data_path_for_alpha(a)
    if not os.path.exists(fpath):
        raise FileNotFoundError(
            f"Dataset mancante per alpha={a} → {fpath}\n"
            "Generalo prima (RST in R) o crea il CSV con una colonna 'y'."
        )
    return pd.read_csv(fpath)["y"].to_numpy()

# Piccola utility per avere un timestamp ad alta risoluzione (per timing).
def now(): 
    return time.perf_counter()

print("=== ST univariata — GRID alpha (Python) — start ===", flush=True)

# Liste che accumulano risultati (righe) e tempi (per ogni alpha).
rows, times = [], []

# Loop principale sulla griglia di alpha.
for a in alpha_grid:
    # (A) Lettura dati per questo valore di alpha
    print(f"\n[alpha={a:+.1f}] Carico dati…", flush=True)
    y = read_y(a)
    n = y.size
    print(f"  → n = {n}", flush=True)

    # (B) Fit ST (stima alpha, nu, xi, omega con skewt.fit)
    print("  Fit skew-t…", flush=True)
    t0 = now()
    # skewt.fit restituisce: a≡alpha, df≡nu, loc≡xi, scale≡omega
    a_hat, df_hat, loc_hat, scale_hat = skewt.fit(y)
    t_fit = now() - t0
    print(
        f"  → stime: alpha={a_hat:.6g}, nu={df_hat:.6g}, xi={loc_hat:.6g}, omega={scale_hat:.6g}  "
        f"(fit {t_fit:.3f}s)",
        flush=True
    )

    # (C) Log-likelihood al massimo
    # Calcola la somma dei log-densità della skew-t stimata, per avere loglik.
    loglik = float(np.sum(skewt.logpdf(y, a=a_hat, df=df_hat, loc=loc_hat, scale=scale_hat)))

    # (D) Inversione CDF↔PPF (core + code)
    # Costruisce una griglia di probabilità:
    # - p_core: parte centrale (0.001–0.999)
    # - p_ext: code estreme (1e-6, 1e-5, 1e-4, 1-1e-4, 1-1e-5, 1-1e-6)
    # Poi:
    #   q_inv  = PPF(p_grid)
    #   p_back = CDF(q_inv)
    # e misura l'errore massimo in core ed estremi separatamente.
    p_core = np.linspace(0.001, 0.999, 999)
    p_ext  = np.array([1e-6, 1e-5, 1e-4, 1-1e-4, 1-1e-5, 1-1e-6])
    p_grid = np.concatenate([p_ext[:3], p_core, p_ext[3:]])
    q_inv  = skewt.ppf(p_grid, a=a_hat, df=df_hat, loc=loc_hat, scale=scale_hat)
    p_back = skewt.cdf(q_inv,   a=a_hat, df=df_hat, loc=loc_hat, scale=scale_hat)

    # Errore massimo nel core (escludendo 3 punti di coda iniziali e finali).
    inv_core = float(np.max(np.abs(p_back[3:-3] - p_core)))
    # Errore massimo nelle code (primi 3 e ultimi 3 punti).
    inv_ext  = float(np.max(np.abs(np.concatenate([p_back[:3], p_back[-3:]]) - p_ext)))

    # (E) UNIF
    # Calcola u = F(y; theta_hat) (PIT) e applica il test KS contro Uniform(0,1).
    u = skewt.cdf(y, a=a_hat, df=df_hat, loc=loc_hat, scale=scale_hat)
    ks_p = float(stats.kstest(u, "uniform", args=(0, 1)).pvalue)

    # (F) Accumulo risultati in memoria
    # row: contiene stime, loglik, misure di inversione e p-value KS per questo alpha.
    rows.append({
        "engine": "Python",
        "alpha_true": a,
        "xi_hat": float(loc_hat),
        "omega_hat": float(scale_hat),
        "alpha_hat": float(a_hat),
        "nu_hat": float(df_hat),
        "loglik": loglik,
        "n": int(n),
        "inv_core_maxerr": inv_core,
        "inv_ext_maxerr": inv_ext,
        "ks_pvalue": ks_p,
    })
    # times: logga solo il tempo di fit per scenario (alpha).
    times.append({"alpha_true": a, "step": "fit_skewt", "seconds": float(t_fit)})

# Converte la lista di dizionari in DataFrame e salva la tabella finale.
pd.DataFrame(rows).to_csv(OUT_TAB, index=False)
pd.DataFrame(times).to_csv(OUT_TIM, index=False)

print(f"\n✓ Scritto: {OUT_TAB}")
print(f"✓ Scritto: {OUT_TIM}")
print("\n=== ST univariata — GRID alpha (Python) — done ===")
